### ISSUE TITLE
Expose ListView velocity when scrolling

### ISSUE THREAD

**Issue by matthiasbruns** [flutter/flutter#26418]
I want to achieve this effect:
![alt text](https://i.stack.imgur.com/wFUEH.gif "Velocity based scrolling")

Here the gif from a google group post, where the dev tries to achieve something similar:
![alt text](https://lh3.googleusercontent.com/-cyFVwkesvZE/W2m4CP3DtlI/AAAAAAAABWU/f743nEULihoL_zJz2YgoGqmZpZBod08uwCLcBGAs/s320/scrolling-list-vertical.gif "Velocity based scrolling")
https://groups.google.com/forum/#!topic/flutter-dev/r0nR6nPCNHw

This shows a list, which item padding changes based on the user scrolling. For this, I need to know, when the user's scrolling velocity is faster or lower than a certain threshold.

I know that I can access ScrollNotifications by wrapping a ListView in a NotificationListener. The notifications store informations like idle, reverse and forward scrolling, which type of scrolling interaction, etc.

Unfortunately, it doesn't seem to store any information about delta scrolling offsets or velocity. I've tried to access the ScrollActivity from the ScrollController attached to the ListView. The activity seems to be protected, which might break my app on release. Here is some code:

```dart
ScrollController _scrollController; // ListView controller
AnimationController _itemSpacingController; // Smooth change of padding for the list items

// Build the list
Widget _buildSectionPage() {
  var section = widget.section;
  var itemCount = section.cells.length;
  return NotificationListener<ScrollNotification>(
    onNotification: _onScroll,
    child: ListView.builder(
      controller: _scrollController,
      itemBuilder: (BuildContext context, int index) {
        return _buildSectionCell(
            context, section.cells[index], index, itemCount);
      },
      itemCount: itemCount,
      padding: EdgeInsets.only(right: 8.0, top: 16.0),
      itemExtent: SectionView._cellHeight +
          (1.0 + _itemSpacingController.value) * DEFAULT_ITEM_SPACING,
      physics: AlwaysScrollableScrollPhysics(),
    ),
  );
}
```

**_buildSectionPage()** sets up the list and hooks up the listeners. They cann _onScroll() when the ListView is being scrolled.

```dart
  bool _onScroll(ScrollNotification notification) {
    if (notification is UserScrollNotification) {
      if (notification.direction == ScrollDirection.idle) {
        _itemSpacingController.reverse();
      } else {
        if (_scrollController.position.activity.velocity <= 0.5) {
          _itemSpacingController.forward();
        }
      }
    }
    return false;
  }
```

> scrollController.position.activity.velocity

is protected. This seems to be the only way to get the current velocity of the scroll event. 
I couldn't find any hint that I could possibly get the velocity from somewhere else.

My ideas

1. What about exposing the delta scroll offset in the ScrollNotifications, because they are being generated from the ScrollPosition, which has access to the activity and its velocity 
2. Expose the velocity in the ScrollController 
3. Add more informations on the listener callbacks of the ScrollController

**Comment by zoechi** [comment]
Please add the output of `flutter doctor -v`.

**Comment by orestesgaolin** [comment]
@matthiasbruns have you achieved your desired effect? I would also be very happy if the velocity of the scroll would be exposed through ScrollController.

**Comment by criswonder** [comment]
@orestesgaolin I also wanna access the scroll velocity when scroll a grid view, it would be great if it can access through ScrollController

**Comment by matthiasbruns** [comment]
Ok I fear I'll have to really do this now :D

**Comment by aaronscherbing** [comment]
In the meantime, I was able to do something like this to workaround my velocity needs.

```
double _pixels;
int _timestamp;

void _scrollListener() {
  double pixels = this._scrollController.position.pixels;
  int timestamp = DateTime.now().millisecondsSinceEpoch;
  if (this._pixels != null) {
    final double velocity = (pixels - this._pixels) / (timestamp - this._timestamp);
    // TODO handle velocity information here.
  }
  this._pixels = pixels;
  this._timestamp = timestamp;
}
```

**Comment by dcbacarro** [comment]
Hi, I am trying to re-create this scroll effect. Any update on this?

**Comment by adrian-alex** [comment]
Hi, please let us know if you have any updates.

**Comment by vladventura** [comment]
Hello everyone :) I made a small adjustment to the scroll position file in my local flutter to help me with this issue, but I don't know if this should be the solution at all because I have little experience with everything :( I'm really sorry!

What i did was expose the velocity only from the activity, like so
`  double get velocity => _activity.velocity;
  double _velocity;
`
It works but like I said, I don't have enough experience to say "this is a good solution", so I need advice on best practices. Thank you!

**Comment by changalberto** [comment]
Is this still being worked on?

**Comment by matthiasbruns** [comment]
let me get to it - I was too busy and forgot the issue completely

**Comment by bugeats** [comment]
As of this date, `ScrollPosition` still does not appear to have a public `velocity` property.

**Comment by dnfield** [comment]
Could you try just using a custom `ScrollPhysics` that overrides the `createBallisticSimulation` so that you can grab the simulation and use it for your positioning?

**Comment by pstromberg98** [comment]
https://github.com/pstromberg98/scroll_velocity_listener
Here is a package for getting velocity change updates on scroll views. It is very simple right now and contributions are welcome.

**Comment by dnfield** [comment]
@pstromberg98 nice. I think it'd be really helpful if you added an example folder to that project with an example that achieves the effect in this bug - or honestly even just a unit test that shows it.

If that's done I think we can safely close this.

**Comment by pstromberg98** [comment]
@dnfield I have added a simple usage example to the readme. I am going to write out some tests and runnable examples when I find time later today

**Comment by pstromberg98** [comment]
While the effect isn't the exact one shown above, I have created a library that adds spring like behavior to scrolling.

Check it out here: https://pub.dev/packages/flutter_bouncy

@dnfield 
I think the two libraries I have provided should be sufficient for handling many use cases including the ones mentioned in this issue, so I think it could be closed. I would add an example (to the scroll_velocity_listener) that achieves the aforementioned effect but the implementation complexity would be too high for an example (I tried simply tying the velocity to padding and it looked terrible). It also seems outside of the scope of scroll_velocity_listener which should only care about calculating scroll velocity.
